### PR TITLE
feat: add backend logic for session requests

### LINKED ISSUE #27: [BE] User - I can request a session on a mentorable position
### Description:
Once a user has mentorable position rights assigned to them, they should be able to submit a request for a session on that position. They should be given an option to make the request recurring, or to make it a one-off-request

### Acceptance Criteria / Requirements:
AC1: If a user has availability, and doesn't have a pending session on that bookable position, they should be able to request a session
AC2: Requests can be made to be one-off (i.e. consumed on acceptance of the session), or can be set to recur (i.e. make the user appear available to mentors to pick up)
AC3: Requests can be deleted

### PR DESCRIPTION
**Pull request by AxonC**
Depends on #66 for changes to documentation etc.

Functionality, as it stands for session requests, include:
- creation
- accepting from a mentors perspective (no permissions just yet)
- revoke (cancelled before accepted)
- cancel

Summary:
- Adds service for bookings in the service container
- Adds table for session requests to be created & managed
- Adds flag in bookings service to forgoe checks which exist traditionally for making a booking which doesn't apply to mentoring sessions (e.g. ratings checks, special endorsements etc.) 
 - Adds SoftDelete to Bookings table & trait included onto the model. This is modelled around the behaviour that when a session request is cancelled, the corresponding booking is also deleted

### CODE REVIEW

File: app/Modules/Bookings/Services/BookingsService.php [hunk 1]
```diff
@@ -132,20 +132,20 @@ public function validateSoloEndorsementEligibility(User $user, Position $positio
         return ! $userAssignedSoloEndorsements->isEmpty();
     }
 
-    public function createBooking(array $bookingData): Booking
+    public function createBooking(array $bookingData, bool $isMentoringSession = false): Booking
```
Comment (atoff) [inline app/Modules/Bookings/Services/BookingsService.php:135]: Give this a docbloc

File: app/Modules/SessionRequest/SessionRequestService.php [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+class SessionRequestService
+{
+    private BookingsService $bookingsService;
+
+    public function __construct(BookingsService $bookingsService)
+    {
+        $this->bookingsService = $bookingsService;
+    }
+
+    public function createSessionRequest(User $user, Position $position)
```
Comment (atoff) [inline app/Modules/SessionRequest/SessionRequestService.php]: docblocs

File: app/Modules/SessionRequest/SessionRequestService.php [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    {
+        $existingSessionRequests = SessionRequest::where('user_id', $user->id)->where(
+            'position_id',
+            $position->id
+        )->get();
+
+        throw_if(! $existingSessionRequests->isEmpty(), SessionRequestAlreadyExistsException::class);
```
Comment (atoff) [inline app/Modules/SessionRequest/SessionRequestService.php:37]: Do we need to consider pilot vs atc session requests here?

File: app/Modules/SessionRequest/SessionRequestService.php [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+        return SessionRequest::create([
+            'user_id' => $user->id,
+            'position_id' => $position->id,
+        ]);
+    }
+
+    public function revokeSessionRequest(SessionRequest $sessionRequest)
+    {
+        throw_if($sessionRequest->isTaken(), SessionRequestAlreadyAcceptedException::class);
```
Comment (atoff) [inline app/Modules/SessionRequest/SessionRequestService.php]: Might be good to be consistent with naming here - do we want to call everything 'accepted' or 'taken'? I would argue accepted is better than taken here

File: composer.json [hunk 5]
```diff
@@ -8,7 +8,7 @@
     ],
     "license": "MIT",
     "require": {
-        "php": "^7.3",
+        "php": "^7.4",
```
Comment (atoff) [inline composer.json:11]: Any specific reason for dropping 7.3?

File: database/factories/SessionRequestFactory.php [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+use App\Modules\Position\Position;
+use App\Modules\SessionRequest\SessionRequest;
+use Faker\Generator as Faker;
+
+$factory->define(SessionRequest::class, function (Faker $faker) {
+    return [
+        'user_id' => 12345,
+        'position_id' => factory(Position::class),
+        'booking_id' => factory(Booking::class),
+        'cancelled_at' => null,
```
Comment (atoff) [inline database/factories/SessionRequestFactory.php:15]: Do we need to explicitly set this?

**Review by atoff — CHANGES_REQUESTED** [review]
Is this targeting #27? - Make sure to link it.

File: app/Modules/SessionRequest/SessionRequestService.php (hunk 3 above)
Comment (AxonC) [reply app/Modules/SessionRequest/SessionRequestService.php:37]: That concept doesn't yet exist, but I suspect we will in the future

File: composer.json (hunk 5 above)
Comment (AxonC) [reply composer.json:11]: typed properties only exist in 7.4

File: app/Modules/SessionRequest/SessionRequestService.php (hunk 4 above)
Comment (AxonC) [reply app/Modules/SessionRequest/SessionRequestService.php]: yeah good point actually, I've done this at the database level too https://github.com/VATSIM-UK/ukts/pull/67/commits/dd3848a8dac67311fec30c932a178a4bf75ff472

File: app/Modules/SessionRequest/SessionRequestService.php (hunk 2 above)
Comment (AxonC) [reply app/Modules/SessionRequest/SessionRequestService.php]: done

File: app/Modules/Bookings/Services/BookingsService.php (hunk 1 above)
Comment (AxonC) [reply app/Modules/Bookings/Services/BookingsService.php:135]: done

File: database/factories/SessionRequestFactory.php (hunk 6 above)
Comment (AxonC) [reply database/factories/SessionRequestFactory.php:15]: I felt it helped to be explicit in the factory, especially if we add states to it in the future

**Comment by AxonC** [comment]
> Is this targeting #27? - Make sure to link it.

Done

File: app/Modules/SessionRequest/SessionRequest.php [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    use SoftDeletes;
+
+    public $guarded = [];
+
+    public function booking(): BelongsTo
+    {
+        return $this->belongsTo(Booking::class);
+    }
+
+    public function isTaken(): bool
```
Comment (atoff) [inline app/Modules/SessionRequest/SessionRequest.php]: We should rename this method as well

**Review by atoff — CHANGES_REQUESTED** [review]

File: app/Modules/SessionRequest/SessionRequest.php (hunk 7 above)
Comment (AxonC) [reply app/Modules/SessionRequest/SessionRequest.php]: good spot, fixed

**Review by atoff — APPROVED** [review]

**Pull request merged**